Sources/Plasma/CoreLib/hsDarwin.h

```diff
@@ -0,0 +1,114 @@
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+Additional permissions under GNU GPL version 3 section 7
+
+If you modify this Program, or any covered work, by linking or
+combining it with any of RAD Game Tools Bink SDK, Autodesk 3ds Max SDK,
+NVIDIA PhysX SDK, Microsoft DirectX SDK, OpenSSL library, Independent
+JPEG Group JPEG library, Microsoft Windows Media SDK, or Apple QuickTime SDK
+(or a modified version of those libraries),
+containing parts covered by the terms of the Bink SDK EULA, 3ds Max EULA,
+PhysX SDK EULA, DirectX SDK EULA, OpenSSL and SSLeay licenses, IJG
+JPEG Library README, Windows Media SDK EULA, or QuickTime SDK EULA, the
+licensors of this Program grant you additional
+permission to convey the resulting work. Corresponding Source for a
+non-source form of such a combination shall include the source code for
+the parts of OpenSSL and IJG JPEG Library used as well as that of the covered
+work.
+
+You can contact Cyan Worlds, Inc. by email legal@cyan.com
+ or by snail mail at:
+      Cyan Worlds, Inc.
+      14617 N Newport Hwy
+      Mead, WA   99021
+
+*==LICENSE==*/
+
+#ifndef _hsDarwin_inc_
+#define _hsDarwin_inc_
+
+#include <string_theory/string>
+#include <string_theory/format>
+
+#ifdef HS_BUILD_FOR_APPLE
+#include <CoreFoundation/CoreFoundation.h>
+
+[[nodiscard]]
+#if __has_feature(attribute_cf_returns_retained)
+__attribute__((cf_returns_retained))
+#endif
+inline CFStringRef CFStringCreateWithSTString(const ST::string& str)
+{
+    return CFStringCreateWithBytes(kCFAllocatorDefault, (const UInt8*)str.data(), str.size(), kCFStringEncodingUTF8, false);
+}
+
+inline ST::string STStringFromCFString(CFStringRef str, ST::utf_validation_t validation = ST_DEFAULT_VALIDATION)
+{
+    CFRange range = CFRangeMake(0, CFStringGetLength(str));
+    CFIndex strBufSz = 0;
+    CFStringGetBytes(str, range, kCFStringEncodingUTF8, 0, false, nullptr, 0, &strBufSz);
+    ST::char_buffer buffer;
+    buffer.allocate(strBufSz);
+    CFStringGetBytes(str, range, kCFStringEncodingUTF8, 0, false, (UInt8*)buffer.data(), strBufSz, nullptr);
+
+    return ST::string(buffer, validation);
+}
+
+inline void format_type(const ST::format_spec &format, ST::format_writer &output, CFStringRef str)
+{
+    ST::char_buffer utf8 = STStringFromCFString(str).to_utf8();
+    ST::format_string(format, output, utf8.data(), utf8.size());
+}
+
+
+#ifdef __OBJC__
+@class NSString;
+
+[[nodiscard]]
+#if __has_feature(attribute_ns_returns_retained)
+__attribute__((ns_returns_retained))
+#endif
+inline NSString* NSStringCreateWithSTString(const ST::string& str)
+{
+#if __has_feature(objc_arc)
+    return (NSString*)CFBridgingRelease(CFStringCreateWithSTString(str));
```

Comment:
I think this is incorrect, but I'm not super sure about all of this. The way I understand it, `CFBridgingRelease` takes in an "owned reference" and returns an "unowned reference". This conflicts with the "Create" function name and "returns retained" attribute, which indicate that the return value is an "owned" reference. I *think* a plain `__bridge` cast would be correct here - that way the "owned" reference stays "owned" as it's returned.

```suggestion
    return (__bridge NSString*)CFStringCreateWithSTString(str);
```

Reply:
I believe this code is correct. It's releasing CoreFoundation's handle to the code while transferring it to ARC.

If the string was CFRelease'd later, a normal bridge would be convention.

Apple's example shows a similar scenario:
https://developer.apple.com/documentation/foundation/1587932-cfbridgingrelease

Reply:
Hmm, I think I'm getting confused by the "release" terminology and the extra inline function here... Am I reading correctly that `CFBridgingRelease` in ARC mode does the same thing as a `__bridge_transfer` cast? I see now that we used `__bridge_transfer` before, but it was then changed to `CFBridgingRelease` and I don't understand why :/

You are right that the ownership needs to be transferred to ARC, so a plain `__bridge` would be incorrect/leaky, yeah. My issue is that we don't want to actually `release` the object at this point, because `NSStringCreateWithSTString` is declared to return a retained object. But perhaps I'm overthinking this - ARC *should* understand what needs to be done based on the "returns retained" attribute combined with a `__bridge_transfer` cast.

Reply:
> Hmm, I think I'm getting confused by the "release" terminology and the extra inline function here... Am I reading correctly that `CFBridgingRelease` in ARC mode does the same thing as a `__bridge_transfer` cast? I see now that we used `__bridge_transfer` before, but it was then changed to `CFBridgingRelease` and I don't understand why :/

I'd have to go back through the history but that might be my fault. I've been doing Swift too long where CoreFoundation is also ARC managed...

I might have thought __bridge_transfer was a bridge that didn't move ownership.

Reply:
> moving it to the ObjC category might result in leaks in the plMessageBox stuff?

Yeah, that would only be possible if we move the category out of plClient and use it everywhere (as suggested below).